tools/wptrunner/wptrunner/environment.py

```diff
@@ -84,18 +84,29 @@ def __init__(self, test_paths, ssl_env, pause_after_test, debug_info, options, e
         self.cache_manager = multiprocessing.Manager()
         self.stash = serve.stash.StashServer()
         self.env_extras = env_extras
+        self.env_extras_cms = None
 
 
     def __enter__(self):
         self.stash.__enter__()
         self.ssl_env.__enter__()
         self.cache_manager.__enter__()
-        for cm in self.env_extras:
-            cm.__enter__(self.options)
-        self.setup_server_logging()
+
         self.config = self.load_config()
+        self.setup_server_logging()
         ports = serve.get_ports(self.config, self.ssl_env)
         self.config = serve.normalise_config(self.config, ports)
+
+        assert self.env_extras_cms == None, (
+            "A TestEnvironment object cannot be nested")
+
+        self.env_extras_cms = []
+
+        for env in self.env_extras:
+            cm = env.use(self.options, self.config)
```

Comment:
I think instead of requiring a class with a use method here, we should make the protocol that `env_extras` is an iterable of callables that return context managers, so this line would be:
`cms = env(self.options, self.config)`

and in `SauceConnect` we would s/`def use`/def __call__`

Reply:
Speaking personally, I very much prefer callables to be named with verbs that
describe the effect of their invocation. It's hard to say what `environment()`
does. I know that `use` is a pretty generic term, but it does give a little
more indication that some resource is being utilized.

That said, my priority is fixing the regression, so I'm happy to oblige.